win32ss/gdi/ntgdi/freetype.c

```diff
@@ -2472,6 +2731,7 @@ IntGetOutlineTextMetrics(PFONTGDI FontGDI,
     PSHARED_FACE SharedFace = FontGDI->SharedFace;
     PSHARED_FACE_CACHE Cache;
     FT_Face Face = SharedFace->Face;
+    BOOL bLocked = IntIsFreeTypeLocked();
```

Comment:
That this `IntGetOutlineTextMetrics` helper needs to know whether a lock is locked, and do stuff in accordance, looks like a hack.
Or said in another way: it's used by another function that already holds the freetype lock, but then, why is that other function doing that and needs also to call IntGetOutlineTextMetrics ?

Reply:
I'm not sure about it.

Reply:
I don't know how to fix it.

Reply:
Is there any convenient lock that can be called from the same thread in multiple times sequentially?

Reply:
https://stackoverflow.com/questions/1988324/how-to-alter-the-recursive-locking-behaviour-of-windows-mutex

Reply:
https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/fast-mutexes-and-guarded-mutexes

> Fast mutexes cannot be acquired recursively. If a thread that is already holding a fast mutex tries to acquire it, that thread will deadlock.

Reply:
https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/introduction-to-mutex-objects

> A thread can acquire ownership of a mutex object that it already owns (recursive ownership),

We have to use `KeInitializeMutex` and `KeReleaseMutex` functions.

Reply:
Why can't we detect deadlock automatically?

Reply:
Could you add the following in the `IntGetOutlineTextMetrics` function:
```c
ASSERT(!bLocked);
```
so that in debug mode it breaks into the debugger if `bLocked` is TRUE (meaning someone upper in the call stack had this freetype lock locked already), and post a full backtrace there so that we can see what's going on.

Reply:
Fixed.